### PR TITLE
feat(utils): add dependency constraint filtering for `RuleTester`

### PR DESCRIPTION
**Pull request by bradzacher**
## PR Checklist

- [x] Addresses an existing open issue: ref #1752
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

While working on #5573 I realised that we need infrastructure that allows us to skip tests depending on dependency versions (i.e. the current ESLint or TS version).

This PR adds that functionality into our extended version of the `RuleTester` class.
You will be able to specify constraints like so:

```ts
ruleTester.run('my-rule', rule, {
  valid: [
    {
      code: 'const x = 1;',
      dependencyConstraints: {
        // enforces the semver ">=7.32.0"
        eslint: '7.32.0',

        // i also added this form which allows the user to fully control the semver.satisfies call
        'some-other-module': {
          type: 'semver',
          version: '~6',
          options: {
            includePrerelease: true,
          },
        },
      },
    },
  ],
  invalid: [
    {
      code: 'const x = 1;',
      errors: [ /* ... */ ],
      dependencyConstraints: {
        eslint: '7.32.0',
        // etc - same as above
      },
    },
  ],
});
```

### CODE REVIEW

File: packages/utils/src/eslint-utils/rule-tester/RuleTester.ts
Comment (bradzacher) [inline packages/utils/src/eslint-utils/rule-tester/RuleTester.ts:52]: note: this is new - I added this specifically to make the class testable

File: packages/utils/src/eslint-utils/rule-tester/dependencyConstraints.ts
Comment (JoshuaKGoldberg) [inline packages/utils/src/eslint-utils/rule-tester/dependencyConstraints.ts:44]: 🤔 I'm not convinced it's worth the extra code & API surface area to have `{ type: 'at-least' }` as an option. Users can get the same behavior quickly from the `string` option or with just a little more effort use the `{ type: 'semver' }`.

Proposal: how about removing `'at-least'` and the `type` field altogether? That way the shape would look like:

```ts
interface SemverVersionConstraint {
  readonly range: string;
  readonly options?: Parameters<typeof semver.satisfies>[2];
}

type VersionConstraint = string | SemverVersionConstraint;
```

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
LGTM in general, very nicely done! I'm pumped to unblock the version-specific issues blocked by #1752.

Just requesting changes on discussing simplifying the API, if we can?

File: packages/utils/src/eslint-utils/rule-tester/RuleTester.ts
Comment (bradzacher) [reply packages/utils/src/eslint-utils/rule-tester/RuleTester.ts:183]: I was purposely avoiding spreading because spreading would mean iterating all tests once before you can iterate again to do your checks.

I felt like gaining perf in exchange for one copied loop was worth the trade-off.

Thinking about it now - one way we could do this without the spread would be to define a custom iterator object. Maybe I'll do that!

File: packages/utils/src/eslint-utils/rule-tester/dependencyConstraints.ts
Comment (bradzacher) [reply packages/utils/src/eslint-utils/rule-tester/dependencyConstraints.ts:44]: Yeah part of me was building a surface that was future-proofed in case we added more. But it's probably over-engineering and not worth the effort.

File: packages/utils/src/eslint-utils/rule-tester/RuleTester.ts
Comment (JoshuaKGoldberg) [reply packages/utils/src/eslint-utils/rule-tester/RuleTester.ts:183]: That is a snazzy solution. Though, how many tests are we defining on average per file, and on average how many files? Is this really going to be a noticeable perf change?

Comment (bradzacher) [reply packages/utils/src/eslint-utils/rule-tester/RuleTester.ts:183]: Some rule trsts do have thousands of tests in them! At the very least a few hundred would be the average.

File: packages/utils/src/eslint-utils/rule-tester/dependencyConstraints.ts
Comment (JoshuaKGoldberg) [inline packages/utils/src/eslint-utils/rule-tester/dependencyConstraints.ts:15]:
```suggestion
   * Passing a string for the value is shorthand for a '>=' constraint
```

File: packages/utils/src/eslint-utils/rule-tester/RuleTester.ts
Comment (JoshuaKGoldberg) [inline packages/utils/src/eslint-utils/rule-tester/RuleTester.ts:221]: Nit:

```suggestion
            return {
              ...test,
              only: satisfiesAllDependencyConstraints(
                test.dependencyConstraints,
              ),
            };
```

**Review by JoshuaKGoldberg — APPROVED** [review]
This is great, thanks @bradzacher! 🚀

**Pull request merged**
